File: 0876-middle-of-the-linked-list/0876-middle-of-the-linked-list.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun middleNode(head: ListNode?): ListNode? {
        var slow = head
        var fast = head
        while (fast?.next != null) {
            slow = slow!!.next
            fast = fast!!.next!!.next
        }
        return slow
    }
}

File: 0876-middle-of-the-linked-list/0876-middle-of-the-linked-list.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn middle_node(mut slow: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut fast = slow.clone();
        
        while fast.is_some() && fast.as_ref()?.next.is_some() {
            fast = fast?.next?.next;
            slow = slow?.next;
        }
        slow
    }
}